pkg/submitter/submitter.go

```diff
@@ -0,0 +1,510 @@
+		),
+	}
+
+	return submitter, submitter.resultCh
+}
+
+func (s *Submitter) Start() error {
+	// Check if the staked account is in dispute.
+	var v []byte
+	var dispute bool
+	for i := 0; i < 10; i++ {
+		if v, dispute = s.IsInDispute(); dispute {
+			continue
+		}
+		select {
+		case <-s.ctx.Done(): // Early exit.
+			return nil
+		default:
+		}
+		time.Sleep(1 * time.Second)
+	}
+	if len(v) == 0 {
+		return errors.New("no status result after 10 attempts. this usually means no connection to the DB")
+	}
+
+	status, _ := hexutil.DecodeBig(string(v))
+	if status.Cmp(big.NewInt(1)) != 0 {
+		return errors.New("miner is not able to mine with current status")
+	}
+
+	for {
+		select {
+		case <-s.ctx.Done():
+			return nil
+		case result := <-s.resultCh:
+			if s.lastSubmitCncl != nil {
+				s.lastSubmitCncl()
+			}
+			var ctx context.Context
+			ctx, s.lastSubmitCncl = context.WithCancel(s.ctx)
+			s.handleSubmit(ctx, result)
+		}
+	}
+}
+
+func (s *Submitter) Stop() {
+	if s.lastSubmitCncl != nil {
+		s.lastSubmitCncl()
+	}
+	s.close()
+	level.Info(s.logger).Log("msg", "submitter shutdown complete")
+}
+
+func (s *Submitter) handleSubmit(ctx context.Context, result *mining.Result) {
+	go func(ctx context.Context, result *mining.Result) {
+		var timeToSubmit context.Context
+		var cncl context.CancelFunc
+		lastSubmit, err := s.lastSubmit()
+		if err != nil {
+			level.Error(s.logger).Log("msg", "checking last submit time", "err", err)
+			timeToSubmit, cncl = context.WithTimeout(ctx, 10*time.Second) // Let's try after 10 seconds.
+		} else if lastSubmit < s.cfg.Mine.MinSubmitPeriod.Duration {
+			level.Debug(s.logger).Log("msg", "min transaction submit threshold hasn't passed", "minSubmitPeriod", s.cfg.Mine.MinSubmitPeriod, "lastSubmit", lastSubmit)
+			timeToSubmit, cncl = context.WithTimeout(ctx, s.cfg.Mine.MinSubmitPeriod.Duration-lastSubmit+10*time.Second) // Adding the 10 seconds penalty to be sure we send the tx in the right time!
+		}
+		defer cncl()
+		select {
+		case <-ctx.Done(): // The context was canceled from the Start loop because new work arrived.
+			return
+		case <-timeToSubmit.Done(): // It is time to submit so start the submit loop
+			for {
+				// There is no new challenge so resend any pending solution.
+				if result == nil {
+					if s.solutionPending == nil {
+						return
+					}
+					result = s.solutionPending
+					var ids []int64
+					for _, id := range s.solutionPending.Work.Challenge.RequestIDs {
+						ids = append(ids, id.Int64())
+					}
+					level.Debug(s.logger).Log("msg", "re-submitting a pending solution", "reqIDs", fmt.Sprintf("%+v", ids))
+				}
+				// Set this solution as pending so that if
+				// any of the checks below fail and will be retried
+				// when there is no new challenge.
+				s.solutionPending = result
+
+				profitPercent, err := s.profit() // Call it regardless of whether we use so that is sets the exposed metrics.
+				if s.cfg.Mine.ProfitThreshold > 0 {
+					if err != nil {
```

Comment:
```suggestion
				profitPercent, err := s.profit() // Call it regardless of whether we use so that is sets the exposed metrics.
				if err != nil {
				.....

				if s.cfg.Mine.ProfitThreshold > 0 {
```